src-java/kilda-model/src/main/java/org/openkilda/adapter/FlowSideAdapter.java

```diff
@@ -45,41 +47,45 @@ public static FlowSideAdapter makeIngressAdapter(Flow flow, FlowPath path) {
      */
     public static FlowSideAdapter makeEgressAdapter(Flow flow, FlowPath path) {
         if (path.getCookie().getDirection() == FlowPathDirection.FORWARD) {
-            return new FlowDestAdapter(flow);
+            return new FlowDestAdapter(flow, path);
         } else {
-            return new FlowSourceAdapter(flow);
+            return new FlowSourceAdapter(flow, path);
         }
     }
 
     /**
      * Determine flow side by switchId and produce corresponding side-adapter.
      */
-    public static FlowSideAdapter makeAdapter(SwitchId switchId, Flow flow) {
+    public static FlowSideAdapter makeAdapter(SwitchId switchId, Flow flow, FlowPath flowPath) {
         if (flow.isOneSwitchFlow()) {
             throw new IllegalArgumentException("Unable to determine flow side for one-switch-flow by switch endpoint");
         } else if (switchId.equals(flow.getSrcSwitchId())) {
-            return new FlowSourceAdapter(flow);
+            return new FlowSourceAdapter(flow, flowPath);
         } else if (switchId.equals(flow.getDestSwitchId())) {
-            return new FlowDestAdapter(flow);
+            return new FlowDestAdapter(flow, flowPath);
         } else {
             throw new IllegalArgumentException(String.format(
                     "Unable to map switch %s on any flow side of %s", switchId, flow));
         }
     }
 
-    protected FlowSideAdapter(Flow flow) {
+    protected FlowSideAdapter(Flow flow, FlowPath flowPath) {
         this.flow = flow;
+        this.flowPath = flowPath;
     }
 
     public abstract FlowEndpoint getEndpoint();
 
-    public abstract boolean isMultiTableSegment();
-
     public abstract boolean isDetectConnectedDevicesLldp();
 
     public abstract boolean isDetectConnectedDevicesArp();
 
     public abstract boolean isPrimaryEgressPath(@NonNull PathId pathId);
 
     public abstract boolean isLooped();
+
+    public boolean isMultiTableSegment() {
+        return getEndpoint().getSwitchId().equals(flowPath.getSrcSwitchId())
```

Comment:
... extremely messy. And it based on the excess `pathPath` field. Bad idea, as for me.